[#5484] Add 'storedir' alias for 'store-dir'

ChShersh opened the pull request:
---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.

I've executed the following command:

```
$ cabal --storedir foo/ new-build cabal-install
```

And got this error:

```
ghc-pkg: foo/ghc-8.4.3/package.db: getDirectoryContents:openDirStream: does
not exist (No such file or directory)
```

Without change in this PR it produces some error regarding parsing CLI arguments.

Maybe I should change some existing test to make sure that `storedir` option won't disappear? But I don't know which one and whether I should do this or not...

phadej commented:
I'm :-1:, we are largely inconsistent with `data-dir` but `subdir`, `source-dir` and `builddir`, `destdir`, `dist-dir`, `output-dir`... so `store-dir` is just one more thing "you have to check" everytime, until it sticks.

phadej commented:
Also a general comment, @ChShersh don't push branches to the main repository. CI's will waste time building both the PR and your branch.

hvr commented:
@ChShersh the error you observed seems like an instance of #5516 - is this deterministically reproducible?

ChShersh commented:
@hvr Yes, I can see this error each time by calling:

```
cabal --storedir foo/ new-build cabal-install
```

But I'm not sure that it's relevant to #5516. In my case directory `foo/` doesn't exist. It's just a dummy argument to ensure that there exist an actual option with name `--storedir`. Though, I can see that the directory `foo/ghc-8.4.3/package.db/` is created after this command. I'm not quite understand what `--storedir` should do...

hvr commented:
@ChShersh do you see this also when you give `--storedir` an absolute path? (we also have this other issue #5485)

23Skidoo commented:
> CI's will waste time building both the PR and your branch.

I think this only affects AppVeyor, since for Travis we have a [branch whitelist](https://github.com/haskell/cabal/blob/master/.travis.yml#L9).

hvr commented:
@23Skidoo shall we use https://www.appveyor.com/docs/branches/#white--and-blacklisting then?

23Skidoo commented:
@hvr Done in 1e9df97d1564db77e8a073d35d18df849177edbc.

23Skidoo commented on cabal-install/Distribution/Client/Setup.hs:
```diff
@@ -1293,9 +1293,9 @@ upgradeCommand = configureCommand {
   }
 
 cleanCommand :: CommandUI CleanFlags
-cleanCommand = Cabal.cleanCommand 
```
In the future, please try to put whitespace changes into a separate commit.

ChShersh replied:
@23Skidoo Just curious. Is there any particular reason to do them in separate commit?

23Skidoo replied:
Makes life easier for the reviewer.

ChShersh replied:
@23Skidoo GitHub supports hiding of whitespace changes. You can add `?w=1` manually to the end of the link to hide whitespace changes. Like this:

* https://github.com/haskell/cabal/pull/5527/files?w=1

Oh you can open GitHub settings

<img width="717" alt="screen shot 2018-08-16 at 9 56 15 am" src="https://user-images.githubusercontent.com/4276606/44182809-1d35af00-a13b-11e8-846b-8e770df439cf.png">

But it means that reviewer still needs to do something to make his own life easier. So it's a tradeoff between who does less work. I'd prefer to do less 🙂 But since you're maintainer, it's okay for me to follow this commit style guide.

ChShersh commented:
@hvr 

>  do you see this also when you give --storedir an absolute path?

No, I don't see this issue. Everything works when I specify absolute path :ok_hand:

phadej commented:
@hvr, note that we can
```
cabal --store-dir=/tmp/asdad --logs-dir=/tmp/asdasd new-build Cabal --dry
```

and what's more surprising, these options aren't in `cabal --help` :(

phadej commented:
There are some `-dir` / `-dirs` flags in http://cabal.readthedocs.io/en/latest/nix-local-build.html#commands

phadej commented:
Sorry for interrupting on somewhat agreed change, I just find it weird. *all* `new-build` dir configuration are without dash...

ChShersh commented:
@hvr @23Skidoo @phadej I've addressed issues from this comment and I've added more aliases to proposed options:

* https://github.com/haskell/cabal/pull/5527#issuecomment-413231581

phadej commented on cabal-install/Distribution/Client/Setup.hs:
```diff
@@ -436,7 +436,7 @@ globalCommand commands = CommandUI {
          (reqArg' "DIR" (\x -> toNubList [x]) fromNubList)
 
-      ,option [] ["logs-dir"]
+      ,option [] ["logsdir", "logs-dir"]
```
I *suspect* that CI fails because the new variants are now first, should we add them last?

hvr replied:
@ChShersh it'd be worth a try; but make it a separate commit in this PR so we can easily revert it if necessary

ChShersh commented:
@hvr @phadej @23Skidoo CI looks much better now. I can't see why that one particular build failed, but otherwise CI results look good.

23Skidoo commented:
Merged, thanks!

23Skidoo commented:
Also cherry-picked into `2.4`.
